mllib/src/main/scala/org/apache/spark/mllib/classification/NaiveBayes.scala

```diff
@@ -113,6 +106,55 @@ class NaiveBayesModel private[mllib] (
     }
   }
 
+  def predictProbabilities(testData: RDD[Vector]): RDD[Map[Double, Double]] = {
+    val bcModel = testData.context.broadcast(this)
+    testData.mapPartitions { iter =>
+      val model = bcModel.value
+      iter.map(model.predictProbabilities)
+    }
+  }
+
+  def predictProbabilities(testData: Vector): Map[Double, Double] = {
+    modelType match {
+      case Multinomial =>
+        val prob = multinomialCalculation(testData)
+        posteriorProbabilities(prob)
+      case Bernoulli =>
+        val prob = bernoulliCalculation(testData)
+        posteriorProbabilities(prob)
+      case _ =>
+        // This should never happen.
+        throw new UnknownError(s"Invalid modelType: $modelType.")
+    }
+  }
+
+  protected[classification] def multinomialCalculation(testData: Vector): DenseVector = {
+    val prob = thetaMatrix.multiply(testData)
+    BLAS.axpy(1.0, piVector, prob)
+    prob
+  }
+
+  protected[classification] def bernoulliCalculation(testData: Vector): DenseVector = {
+    testData.foreachActive { (index, value) =>
+      if (value != 0.0 && value != 1.0) {
+        throw new SparkException(
+          s"Bernoulli naive Bayes requires 0 or 1 feature values but found $testData.")
+      }
+    }
+    val prob = thetaMinusNegTheta.get.multiply(testData)
+    BLAS.axpy(1.0, piVector, prob)
+    BLAS.axpy(1.0, negThetaSum.get, prob)
+    prob
+  }
+
+  protected[classification] def posteriorProbabilities(prob: DenseVector): Map[Double, Double] = {
+    val maxLogs = max(prob.toBreeze)
+    val minLogs = min(prob.toBreeze)
+    val normalized = prob.toArray.map(e => (e - minLogs) / (maxLogs - minLogs))
```

Comment:
This is on the right path except that the 'normalization' upfront here isn't valid. Just drop that. I meant you have to subtract the sum of the log of probabilities rather than of log-probabilities, but then I think the easiest thing is just to go straight back to probabilities, then normalize. This introduces some issues with very small probabilities, but I assume the API must return probabilities and not log-probabilities, so this is going to be an issue no matter what. I think this might be about right:

```
val probabilities = prob.toArray.map(math.exp)
val probSum = probabilities.sum
labels.zip(probabilities.map(_ / probSum)).toMap
```